ath79: add support for ZTE MF282

andyboeh opened the pull request:
The ZTE MF282 is a LTE router used (exclusively?) by the network operator "3".

Specifications
==============

SoC: QCA9563 (775MHz)
RAM: 128MiB
Flash: 8MiB SPI-NOR + 128MiB SPI-NAND
LAN: 1x GBit LAN
LTE: ZTE MF270 (Cat4), detected as P685M
WiFi: QCA9880ac + QCA9560bgn

MAC addresses
=============

LAN: from config
WiFi 1: from config
WiFi 2: +1

Installation
============

TFTP installation using UART is preferred. Disassemble the device and connect serial. Put the initramfs image as openwrt.bin to your TFTP server and configure a static IP of 192.168.1.100. Load the initramfs image by typing:

  setenv serverip 192.168.1.100
  setenv ipaddr 192.168.1.1
  tftpboot 0x82000000 openwrt.bin
  bootm 0x82000000

From this intiramfs boot you can take a backup of the currently installed partitions as no vendor firmware is available for download.

Once booted, transfer the sysupgrade image and run sysupgrade.

LTE Modem
=========

The LTE modem is probably the same as in the MF283+, all instructions apply.

Configuring the connection using modemmanager works properly, the modem provides three serial ports and a QMI CDC ethernet interface.

andyboeh committed:
ath79: refactor ZTE MF28x dts files

Move common dts entries of ZTE MF281 and ZTE MF286 to a common .dtsi file
to reduce redundancies.

Signed-off-by: Andreas Böhler <dev@aboehler.at>

Leo-PL commented:
Looks fine at the first glance. I'm not sure how the maintainers would look at the `/delete-node/` and `/delete-property` statements - they will certainly work, but I think that extracting a common `qca9563_zte_mf28x.dtsi` and keeping the specifics for MF286 series in qca9563_zte_mf286.dtsi including the previous one would look a bit cleaner, IMO.

andyboeh commented:
There are several other devices with `/delete-node/` and `/delete-property/` statements, e.g. the FRITZ!7530 .dts. For the ath79 target, I see this in `qca9531_glinet_gl-ar300m16.dts` and `qca9531_glinet_gl-ar300m-lite.dts`.

This is certainly something to discuss, but there is then a bit more work to do since there is a lot of repetition in the 281 dts as well. I didn't want to touch all this as I only have the 282 for testing.

Leo-PL commented:
Okay, understood. I have MF286 and MF286A for testing, so I could check that for you, if you'd like.
Either way:
Reviewed-by: Lech Perczak <lech.perczak@gmail.com>

hauke commented:
I would prefer if the delete-node is not needed.
Maybe you could remove the entries from the `qca9563_zte_mf286.dtsi` file and move them to the two dts files including this or create a new sub dtsi file.

andyboeh commented:
OK, let's do it properly then: I will refactor all ZTE MF28x .dts files in order to reduce repetition.

andyboeh commented:
@hauke Done, but please keep in mind that I only compile-tested the changes to the MF281 and MF286 dts files. This now affects all ZTE MF28x devices.

Leo-PL commented:
Please give me some time to build and test on MF286 and MF286A, then.

Leo-PL commented:
I compared compiled device trees before and after this patch. It seems, that MF281 is now missing `/leds/led-0/color` and `/leds/led0/label` properties. They should be restored - values are the same as in MF286. Others - for MF286* are unchanged, so aside the change needed for MF281, +1 from my side.

Leo-PL commented on target/linux/ath79/dts/qca9563_zte_mf282.dts:
```diff
@@ -0,0 +1,129 @@
+	};
+};
+
+&led_debug {
```
This needs adding to MF281 device tree, with blue changed to green.

andyboeh replied:
Thanks, fixed.

andyboeh commented:
> I compared compiled device trees before and after this patch. It seems, that MF281 is now missing `/leds/led-0/color` and `/leds/led-0/label` properties. They should be restored - values are the same as in MF286. Others - for MF286* are unchanged, so aside the change needed for MF281, +1 from my side.

Thanks for the review, it's easy to miss this if you don't look closely.

Leo-PL commented on target/linux/ath79/dts/qca9563_zte_mf281.dts:
```diff
@@ -3,246 +3,157 @@
 
-	keys {
-		compatible = "gpio-keys";
+&led_debug {
```
You forgot to change colour to green here, as it was set before refactoring.

Leo-PL commented:
Ok, now it's looking fine,
+1.

Leo-PL commented:
@andyboeh I have a suggestion - you can now configure QMI interface from `ucidef` as well, which means working WAN on the modem on first boot, for example:
```
ucidef_set_interface "wan" device "/dev/cdc-wdm0" protocol "qmi"
```
Works for me with MF28D, I'm preparing changes for MF286 series and MF283+ now.

andyboeh commented:
Thanks for the hint. I had this already in, but I couldn't get the modem to work with uqmi (it connects, but no data is transferred). With ModemManager, everything is fine so I decided against this.

Leo-PL commented:
That's strange, for me this worked out of box in MF283+, it's the same modem indeed. ModemManager does some kind of magic with profiles, probably, so you don't have to configure matching APN with `AT+CGDCONT` command.
I would try to configure matching profile with `AT+CGDCONT` and retry with `uqmi`. Bonus: quicker network and LTE attach if there is a proper default bearer configured.

hauke commented:
Could you split this into two commits please, one doing the refactoring and one adding the new device. 
Do you plan to add the `ucidef_set_interface` entry for the wwan?

andyboeh committed:
ath79: add support for ZTE MF282

The ZTE MF282 is a LTE router used (exclusively?) by the network operator
"3".

Specifications
==============

SoC: QCA9563 (775MHz)
RAM: 128MiB
Flash: 8MiB SPI-NOR + 128MiB SPI-NAND
LAN: 1x GBit LAN
LTE: ZTE MF270 (Cat4), detected as P685M
WiFi: QCA9880ac + QCA9560bgn

MAC addresses
=============

LAN: from config
WiFi 1: from config
WiFi 2: +1

Installation
============

TFTP installation using UART is preferred. Disassemble the device and
connect serial. Put the initramfs image as openwrt.bin to your TFTP server
and configure a static IP of 192.168.1.100. Load the initramfs image by
typing:

  setenv serverip 192.168.1.100
  setenv ipaddr 192.168.1.1
  tftpboot 0x82000000 openwrt.bin
  bootm 0x82000000

From this intiramfs boot you can take a backup of the currently installed
partitions as no vendor firmware is available for download.

Once booted, transfer the sysupgrade image and run sysupgrade.

LTE Modem
=========

The LTE modem is probably the same as in the MF283+, all instructions
apply.

Configuring the connection using modemmanager works properly, the modem
provides three serial ports and a QMI CDC ethernet interface.

Signed-off-by: Andreas Böhler <dev@aboehler.at>

andyboeh commented:
Done, split it into two commits and added the `ucidef_set_interface` entry. This works, i.e. the interface is correctly set up and the device gets an IP address.
